lib/view_component/compiler.rb

```diff
@@ -79,11 +81,13 @@ def define_render_template_for
 
       component_class.class_eval <<-RUBY, __FILE__, __LINE__ + 1
         def render_template_for(variant = nil)
-          if variant.nil?
-            call
-          #{variant_elsifs}
-          else
-            call
+          self.class.lock.synchronize do
```

Comment:
We likely don't want to use the mutex in the render path for performance reasons, especially since this is usually only an issue in the test environment.

Happy to take a look later, but I'm curious what breaks without this code.

Reply:
Without adding the thread-safety in render, the build was failing, see [here](https://github.com/github/view_component/runs/4299758830?check_suite_focus=true). And I believe it's right because, if we only add the lock acquisition when we redefine `call`, then, while the code which redefines `call` is running, we can still have `render` calls being made and those can fail.

If performance is an issue, would maybe moving the compilation from render time to loading time be a better idea?

Reply:
> If performance is an issue, would maybe moving the compilation from render time to loading time be a better idea?

I had a similar thought and have been thinking on some of the trade-offs. The big one that comes to mind is that we'd need to somehow determine which components to compile, otherwise we may unnecessarily slow down test suite boot time. That wouldn't be great for cases where you want to run a single test or subset of tests.

Reply:
> > If performance is an issue, would maybe moving the compilation from render time to loading time be a better idea?
> 
> I had a similar thought and have been thinking on some of the trade-offs. The big one that comes to mind is that we'd need to somehow determine which components to compile, otherwise we may unnecessarily slow down test suite boot time. That wouldn't be great for cases where you want to run a single test or subset of tests.

Makes sense...

I see that people start to mention issues appearing in `development` mode as well. Does it make sense to put this locking mechanism behind a check for `env == test || development`?

Reply:
Also, the reason why I switched from `Mutex` to `Monitor` is [this](https://github.com/github/view_component/runs/4300334626?check_suite_focus=true) failing build.

Reply:
Using a `Monitor` makes sense to me. 👍 

I just made a few suggestions in the latest review round. I don't think it will make things 100% thread safe, but should minimize the risk and at least allow us to remove the `synchronize` here.

Similar to your env idea, I'm wondering if we need to go a level higher and think about introducing different modes to the compiler. A consistent compiler, for dev + test, and a fast compiler, for production. I don't think that idea should block this PR, but we do want to keep the performance benefits we see in production while also making dev+test consistent in threaded environments.